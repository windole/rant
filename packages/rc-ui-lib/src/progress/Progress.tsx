import React, { CSSProperties, useContext, useMemo } from 'react';
import classnames from 'classnames';
import { ProgressProps } from './PropsType';
import { addUnit } from '../utils';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

const Progress: React.FC<ProgressProps> = (props) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('progress', prefixCls);
  const { inactive, color, trackColor, percentage, strokeWidth } = props;

  const background = useMemo(() => {
    return inactive ? undefined : color;
  }, [inactive, color]);

  const renderPivot = () => {
    const { textColor, pivotText, pivotColor } = props;
    const text = pivotText ?? `${percentage}%`;

    if (props.showPivot && text) {
      const style = {
        color: textColor,
        left: `${+percentage}%`,
        transform: `translate(-${+percentage}%,-50%)`,
        background: pivotColor || background,
      };

      return (
        <span style={style} className={bem('pivot', { inactive: props.inactive }) as string}>
          {text}
        </span>
      );
    }
    return null;
  };

  const rootStyle: CSSProperties = {
    background: trackColor,
    height: addUnit(strokeWidth),
  };
  const portionStyle: CSSProperties = {
    width: `${percentage}%`,
    background,
  };
  return (
    <div className={classnames(props.className, bem())} style={rootStyle}>
      <span
        className={classnames(
          bem('portion', {
            inactive: props.inactive,
          }),
        )}
        style={portionStyle}
      />
      {renderPivot()}
    </div>
  );
};

Progress.defaultProps = {
  percentage: 0,
  strokeWidth: '4px',
  color: '#1989fa',
  trackColor: '#e5e5e5',
  pivotColor: '',
  textColor: 'white',
  inactive: false,
  showPivot: true,
};

export default Progress;
