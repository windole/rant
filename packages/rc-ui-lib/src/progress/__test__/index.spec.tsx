import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Progress, { ProgressProps } from '..';

describe('Progress', () => {
  function createProgress(props: ProgressProps) {
    const { queryByTestId, container, rerender, debug, baseElement } = render(
      <Progress {...props} />,
    );

    return {
      container,
      rerender,
      debug,
      queryByTestId,
      baseElement,
    };
  }

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  it('should re-calc width if showing pivot dynamically', () => {
    const { rerender, container } = createProgress({
      showPivot: false,
      percentage: 100,
    });
    expect(container).toMatchSnapshot();
    const props = {
      percentage: 100,
      showPivot: true,
      pivotText: 'test',
    };
    rerender(<Progress {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should change track color when using track-color prop', () => {
    const { baseElement } = createProgress({
      showPivot: false,
      percentage: 100,
    });

    expect(getComputedStyle(baseElement.querySelector('.rc-progress')).background).toEqual(
      'rgb(229, 229, 229)',
    );
  });

  it('should render correctly when using inactive prop', () => {
    const { container } = createProgress({
      inactive: true,
      percentage: 50,
    });
    expect(container).toMatchSnapshot();
  });
});
