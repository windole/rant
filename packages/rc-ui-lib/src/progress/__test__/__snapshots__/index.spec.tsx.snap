// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress should re-calc width if showing pivot dynamically 1`] = `
<div>
  <div
    class="rc-progress"
    style="background: rgb(229, 229, 229); height: 4px;"
  >
    <span
      class="rc-progress__portion"
      style="width: 100%; background: rgb(25, 137, 250);"
    />
  </div>
</div>
`;

exports[`Progress should re-calc width if showing pivot dynamically 2`] = `
<div>
  <div
    class="rc-progress"
    style="background: rgb(229, 229, 229); height: 4px;"
  >
    <span
      class="rc-progress__portion"
      style="width: 100%; background: rgb(25, 137, 250);"
    />
    <span
      class="rc-progress__pivot"
      style="color: white; left: 100%; transform: translate(-100%,-50%); background: rgb(25, 137, 250);"
    >
      test
    </span>
  </div>
</div>
`;

exports[`Progress should render correctly when using inactive prop 1`] = `
<div>
  <div
    class="rc-progress"
    style="background: rgb(229, 229, 229); height: 4px;"
  >
    <span
      class="rc-progress__portion rc-progress__portion--inactive"
      style="width: 50%;"
    />
    <span
      class="rc-progress__pivot rc-progress__pivot--inactive"
      style="color: white; left: 50%; transform: translate(-50%,-50%);"
    >
      50%
    </span>
  </div>
</div>
`;
